pkg/store/io.go

```diff
@@ -0,0 +1,123 @@
+package store
+
+import (
+	"bufio"
+	"io"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	readerBufferSize = 32 * 1024
+)
+
+// byteRange holds information about a single byte range.
+type byteRange struct {
+	offset int
+	length int
+}
+
+// byteRanges holds a list of non-overlapping byte ranges sorted by offset.
+type byteRanges []byteRange
+
+// size returns the total number of bytes in the byte ranges.
+func (r byteRanges) size() int {
+	size := 0
+	for _, c := range r {
+		size += c.length
+	}
+	return size
+}
+
+// contiguous returns whether all byte ranges are contiguous (no gaps).
+func (r byteRanges) contiguous() bool {
+	if len(r) < 2 {
+		return true
+	}
+
+	for off, idx := r[0].offset+r[0].length, 1; idx < len(r); idx++ {
+		if r[idx].offset != off {
+			return false
+		}
+		off += r[idx].length
+	}
+	return true
+}
+
+// readByteRanges reads the provided byteRanges from src and append them to dst. The provided
+// byteRanges must be sorted by offset and non overlapping. The byteRanges offset must be
+// relative to the beginning of the provided src (offset 0 == first byte will be read from src).
+func readByteRanges(src io.Reader, dst []byte, byteRanges byteRanges) ([]byte, error) {
+	if len(byteRanges) == 0 {
+		return nil, nil
+	}
+
+	// Ensure the provided dst buffer has enough capacity.
+	expectedSize := byteRanges.size()
+	if cap(dst) < expectedSize {
+		return nil, io.ErrShortBuffer
+	}
+
+	// Size the destination buffer accordingly.
+	dst = dst[0:expectedSize]
+
+	// Optimisation for the case all ranges are contiguous.
+	if byteRanges[0].offset == 0 && byteRanges.contiguous() {
+		// We get an ErrUnexpectedEOF if EOF is reached after some but not all bytes
```

Comment:
Sad but true. I have plans to work on it on a follow up PR, but this PR shouldn't change the current behaviour. The logic in `master` does `_, err = io.Copy(buf, r)` ([see here](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go#L1485)) so doesn't check if the number of read bytes is the expected one (and wouldn't be even possible because we don't have a way to know what's the actual expected one, which is something I would like to fix).

Reply:
I mean a different issue.

The partitioner builds the parts like this:
```
parts := r.block.partitioner.Partition(len(offsets), func(i int) (start, end uint64) {
    return uint64(offsets[i]), uint64(offsets[i]) + maxChunkSize
})
```

Since we don't know the chunk length, we always consider each chunk having `maxChunkSize`. However, when you read the last chunk of the segment file and the chunk size is less than `maxChunkSize` then the expected read bytes (eg. `dst` buffer size) is bigger than the actual size we'll read.

The trick here handle this special case.